Make song select ensure current beatmap is always playable in the active ruleset

naoey opened the pull request:
Alternative to #1830 and fixes #1724 and #2055

naoey committed:
Make song select ensure current beatmap is always playable in the active ruleset.

 - Add a to TestCasePlaySongSelect testing this scenario

naoey committed:
Merge remote-tracking branch 'upstream/master' into fix-unplayable-beatmaps

naoey committed:
More specific tests.

peppy committed:
Merge branch 'master' into fix-unplayable-beatmaps

peppy commented on osu.Game/Screens/Select/SongSelect.cs:
```diff
@@ -194,6 +199,8 @@ private void load(BeatmapManager beatmaps, AudioManager audio, DialogOverlay dia
             if (this.beatmaps == null)
                 this.beatmaps = beatmaps;
 
+            rulesetConversionAllowed = config.GetBindable<bool>(OsuSetting.ShowConvertedBeatmaps);
```
i think we want to bind `rulesetConversionAllowed.ValueChanged += ensurePlayableRuleset` too, right?

naoey replied:
the carousel itself seemed to handle it by changing filters

peppy commented on osu.Game.Tests/Visual/TestCasePlaySongSelect.cs:
```diff
@@ -147,8 +233,9 @@ private BeatmapSetInfo createTestBeatmapSet(int i)
                         OnlineBeatmapID = 1235 + i,
-                        Ruleset = rulesets.AvailableRulesets.First(),
-                        Path = "hard.osu",
+                        Ruleset = rulesets.AvailableRulesets.First(r => r.ID != 0),
```
probably use ElementAt() here too

peppy commented on osu.Game/Screens/Select/SongSelect.cs:
```diff
@@ -443,6 +456,34 @@ private void ensurePlayingSelected(bool preview = false)
             }
         }
 
+        private void ensurePlayableRuleset()
```
Is there any reason for this to be here rather than in `BeatmapCarousel`? You're duplicating the "can we convert" conditional out of the carousel scope (and doing so twice in the same method). Need to avoid duplicating such code wherever possible.

I'd suggest trying to work this in to `BeatmapCarousel`'s `SelectBeatmap`, which is already called in the necessary code paths.

Then you can probably already check if the beatmap to be selected is currently filtered, and if so, perform the required actions. Also I'm not sure whether we want it to trigger a ruleset change, maybe leave that part out (and call `SelectNextRandom` instead).

naoey replied:
the ruleset change was to handle the scenario where the next track in the music controller might be an incompatible ruleset. when pressing "next" in music controller loads a random beatmap instead it'd be a bit disconcerting (like in #2055)

peppy replied:
I think the more common scenario is the one we should cater to initially – entering song select from the main menu after a random beatmap starts playing. We don't want this to switch a user's active ruleset.

naoey committed:
Remove previous fix and move filtered logic to carousel.

- Add an optional bool parameter to SelectBeatmap to skip selecting
filtered maps

naoey committed:
Make song select choose random when initial selection fails.

- Revert TestCasePlaySongSelect to master

peppy committed:
Merge branch 'master' into fix-unplayable-beatmaps

naoey committed:
Invert bool, add test, and handle ruleset change.

naoey committed:
Merge branch 'fix-unplayable-beatmaps' of github.com:naoey/osu into fix-unplayable-beatmaps

peppy committed:
Merge branch 'master' into fix-unplayable-beatmaps

peppy committed:
Use private instead of protected
